### DISCUSSION TITLE
Forms replacement for REST API

### DISCUSSION THREAD

**Discussion by xorik** [symfony/symfony#43191] [Q&A]
What are you using in your APIs, when you don't need HTML?

I'm stick with forms, but experience a lot of troubles. Syntax is really bad, no typing and hard to read and explain to junior devs.

In other hands it saves lot of time and code: I don't need manually call setters and validate every field manually. Also I can easily replace ID with an actual object from database.

Any suggestions?

**Comment by norkunas** [comment]
You could try https://api-platform.com/ :)

**Reply by xorik** [reply]
Thanks, I may consider it for my future projects. But I'm not sure if it's worth rewriting existing code with it..

**Comment by michaoj** [comment]
I use the DTO approcah. Like this you can validate your DTO and once it is correct persists it using your entity object.

**Reply by xorik** [reply]
I was thinking about it. But I can see some issues:
- I need manually call setters on my model
- There is no way automatically convert string ID to object, so I need call a repository method manually
- Not sure how to make/validate nested DTOs

Also I need to invent some tool to collect all validation/type issues and prepare a clear response to frontend

**Reply by michaoj** [reply]
You can overcome those 3 points : 

- Your DTO's can inherit an abstract one which implements a generic method to load your object based on your DTO
- You can use a param converter or your own validation rule to convert on the fly id to object
- Nested validation shouldn't be a problem (see https://github.com/symfony/symfony/discussions/43184)

And of course as @norkunas stated, you can try apiplatform ;)

**Reply by xorik** [reply]
Thanks, looks interesting! Especially I didn't know that the paramconverter works with validation.

About the api platform: I've checked briefly their docs, but didn't find any example related to DTO validation/mapping. Could you point me to an example, maybe I can easily reuse this logic, without rewriting the app.

**Reply by norkunas** [reply]
https://api-platform.com/docs/core/dto/

**Reply by michaoj** [reply]
My bad, the param converter only applies on a Request object
